modules/ocha_ai_tag/src/Services/OchaAiTagTagger.php

```diff
@@ -0,0 +1,348 @@
+   *
+   * @param array $texts
+   *   Texts to embed.
+   * @param bool $query
+   *   TRUE if the text is a search query.
+   *
+   * @return array
+   *   Embedding for the text.
+   *
+   * @throws \Exception
+   *   Throw an exception if the generation of the embeddding fails.
+   */
+  protected function requestEmbeddings(array $texts, bool $query = FALSE): array {
+    if (empty($texts)) {
+      return [];
+    }
+
+    $data = $this->getEmbeddingPlugin()->generateEmbeddings($texts, $query);
+
+    return $data;
+  }
+
+  /**
+   * Get the most probable terms for the text.
+   *
+   * @param array $embeddings
+   *   Embeddings of a text as associative arrays with text and embedding
+   *   properties.
+   * @param bool $average_embeddings
+   *   If TRUE, generate the mean of the embeddings and compute the similarities
+   *   against it.
+   * @param array $types
+   *   Types of formula to use to compare the similarity scores. One of "max",
+   *   "mean" or "mean_with_cutoff" or "max_mean".
+   * @param float|null $alpha
+   *   Coefficient to calculate the cutoff value.
+   *
+   * @return array
+   *   Associative array with vacobularies as keys and list of terms and their
+   *   similarity score as values.
+   */
+  protected function getSimilarTerms(array $embeddings, bool $average_embeddings = FALSE, array $types = [self::CALCULATION_METHOD_MAX], $alpha = 0.2): array {
+    if ($average_embeddings) {
+      $embeddings = [
+        VectorHelper::mean($embeddings),
+      ];
+    };
+
+    $vocabularies = [];
+    foreach ($this->getTermEmbeddings() as $vocabulary => $term_embeddings) {
+      $results = [];
+      foreach ($term_embeddings as $term => $term_embedding) {
+        $similarities = [];
+
+        foreach ($embeddings as $item) {
+          $similarities[] = VectorHelper::cosineSimilarity($term_embedding, $item);
+        }
+
+        foreach ($types as $type) {
+          switch ($type) {
+            case 'max':
+              $results[$type][$term] = max($similarities);
+              break;
+
+            case 'mean':
+              $results[$type][$term] = array_sum($similarities) / count($similarities);
+              break;
+
+            case 'mean_with_cutoff':
+              $similarities = $this->filterSimilarities($similarities, $alpha);
+              $results[$type][$term] = array_sum($similarities) / count($similarities);
+              break;
+
+            // Max x mean.
+            default:
+              // Max x mean.
+              $results[$type][$term] = max($similarities) * array_sum($similarities) / count($similarities);
+          }
+        }
+      }
+
+      // Sort by similarity score descending.
+      foreach ($types as $type) {
+        arsort($results[$type]);
+        $vocabularies[$type][$vocabulary] = $results[$type];
+      }
+    }
+    return $vocabularies;
+  }
+
+  /**
+   * Get the embeddings for the taxonomy terms.
+   */
+  protected function getTermEmbeddings(): array {
```

Comment:
Another thing is that there is currently no way to clear the stored embeddings. That means the embeddings will not change even if the term names or descriptions are changed (better term descriptions can be a good way to improve the similarity comparison).